### PR TITLE
build: new `ddev mariadb` command, adjust `~/.my.cnf` to always use `root`, hide some Docker build warnings, fixes #6462

### LINKED ISSUE #6462: Review docker builds for warnings and errors
We have some warnings in Docker image builds. It would be great to chase these and resolve them. Some are a result of `docker buildx build` tightening up. I think there may be one from `docker-compose build` at start time as well.

We should review the build logs and address these.

### PR DESCRIPTION
**Pull request by stasadev**
## The Issue

- #6462 

## How This PR Solves The Issue

We can skip the annoying warning `SecretsUsedInArgOrEnv` that we can't get rid of.

- https://stackoverflow.com/a/79156231/8097891
- https://docs.docker.com/build/checks/#skip-checks

The `~/.my.cnf` in the `web` image is the same as in the `db` image and has `pager=`.

`ddev mysql` prefers `mariadb` when it's available.

A new `ddev mariadb` command that works the same as `ddev mysql`, but only for the MariaDB type. It is only needed because we don't have autocomplete with alias, and MariaDB is promoting its own commands now, we shouldn't hide it under `ddev mysql`.

Add `/usr/local/mysql/bin` for `mysql:5.5` in $PATH inside the `db` container.

## Manual Testing Instructions

```
ddev config --database=mysql:8.0 && ddev start

# it should have pager:
ddev mysql -e "SELECT * FROM INFORMATION_SCHEMA.TABLES;"
```

```
ddev delete -Oy
ddev config --database=mariadb:11.4 && ddev start

# it should have pager, and no warning about deprecated name
ddev mysql -e "SELECT * FROM INFORMATION_SCHEMA.TABLES;"
ddev mariadb -e "SELECT * FROM INFORMATION_SCHEMA.TABLES;"
```

```
ddev delete -Oy
ddev config --database=mysql:5.5 && ddev start

ddev ssh -s db
# the command should work (it didn't work before: mysql: command not found)
mysql -V
```

### CODE REVIEW

File: containers/ddev-dbserver/Dockerfile [hunk 1]
```diff
@@ -1,3 +1,6 @@
+# syntax=docker/dockerfile:1
+# check=skip=SecretsUsedInArgOrEnv
```
Comment (rfay) [inline containers/ddev-dbserver/Dockerfile:2]: Nice!

**Review by rfay — APPROVED** [review]
Yay! Now if we figure out how to quiet the mysql client complaint about password used on command line :)

**Comment by stasadev** [comment]
I experimented with mariadb warning:

```
$ ddev exec -s db mysql -V
mysql: Deprecated program name. It will be removed in a future release, use '/usr/bin/mariadb' instead
mysql from 11.4.4-MariaDB, client 15.2 for debian-linux-gnu (x86_64) using  EditLine wrapper
```

MariaDB looks on the executable name, and adds a warning if the command doesn't start with `mariadb*`:
```
$ ddev exec -s db "ln -s /usr/bin/mariadb mmariadb"
$ ddev exec -s db "./mmariadb -V"
./mmariadb: Deprecated program name. It will be removed in a future release, use '/usr/bin/mariadb' instead
./mmariadb from 11.4.4-MariaDB, client 15.2 for debian-linux-gnu (x86_64) using  EditLine wrapper
```
So it doesn't really matter if we have a symlink or not.

I think calling `mariadb*` from bash script wrappers can break command arguments.

I tried to wrap it in functions and aliases.

Aliases work only inside the db container, not with `ddev exec`. Functions work everywhere (if you `source` them), but they break `pid="$!"` functionality, when you run `mysqld` and want to get the pid.

In the end, I decided that none of them were worth adding.

```bash
# Command name changes for MariaDB 11+
# Only works inside the db container

if command -v mariadb >/dev/null 2>&1; then
    alias mysql="mariadb"
fi

if command -v mariadb-access >/dev/null 2>&1; then
    alias mysqlaccess="mariadb-access"
fi

if command -v mariadb-admin >/dev/null 2>&1; then
    alias mysqladmin="mariadb-admin"
fi

if command -v mariadb-backup >/dev/null 2>&1; then
    alias mariabackup="mariadb-backup"
fi

if command -v mariadb-binlog >/dev/null 2>&1; then
    alias mysqlbinlog="mariadb-binlog"
fi

if command -v mariadb-check >/dev/null 2>&1; then
    alias mysqlanalyze="mariadb-check"
    alias mysqlcheck="mariadb-check"
    alias mysqloptimize="mariadb-check"
    alias mysqlrepair="mariadb-check"
fi

if command -v mariadb-convert-table-format >/dev/null 2>&1; then
    alias mysql_convert_table_format="mariadb-convert-table-format"
fi

if command -v mariadb-dump >/dev/null 2>&1; then
    alias mysqldump="mariadb-dump"
fi

if command -v mariadb-dumpslow >/dev/null 2>&1; then
    alias mysqldumpslow="mariadb-dumpslow"
fi

if command -v mariadb-find-rows >/dev/null 2>&1; then
    alias mysql_find_rows="mariadb-find-rows"
fi

if command -v mariadb-fix-extensions >/dev/null 2>&1; then
    alias mysql_fix_extensions="mariadb-fix-extensions"
fi

if command -v mariadb-hotcopy >/dev/null 2>&1; then
    alias mysqlhotcopy="mariadb-hotcopy"
fi

if command -v mariadb-import >/dev/null 2>&1; then
    alias mysqlimport="mariadb-import"
fi

if command -v mariadb-install-db >/dev/null 2>&1; then
    alias mysql_install_db="mariadb-install-db"
fi

if command -v mariadb-plugin >/dev/null 2>&1; then
    alias mysql_plugin="mariadb-plugin"
fi

if command -v mariadb-report >/dev/null 2>&1; then
    alias mysqlreport="mariadb-report"
fi

if command -v mariadb-secure-installation >/dev/null 2>&1; then
    alias mysql_secure_installation="mariadb-secure-installation"
fi

if command -v mariadb-setpermission >/dev/null 2>&1; then
    alias mysql_setpermission="mariadb-setpermission"
fi

if command -v mariadb-show >/dev/null 2>&1; then
    alias mysqlshow="mariadb-show"
fi

if command -v mariadb-slap >/dev/null 2>&1; then
    alias mysqlslap="mariadb-slap"
fi

if command -v mariadb-tzinfo-to-sql >/dev/null 2>&1; then
    alias mysql_tzinfo_to_sql="mariadb-tzinfo-to-sql"
fi

if command -v mariadb-upgrade >/dev/null 2>&1; then
    alias mysql_upgrade="mariadb-upgrade"
fi

if command -v mariadb-waitpid >/dev/null 2>&1; then
    alias mysql_waitpid="mariadb-waitpid"
fi

if command -v mariadbd >/dev/null 2>&1; then
    alias mysqld="mariadbd"
fi

if command -v mariadbd-multi >/dev/null 2>&1; then
    alias mysqld_multi="mariadbd-multi"
fi

if command -v mariadbd-safe >/dev/null 2>&1; then
    alias mysqld_safe="mariadbd-safe"
fi

if command -v mariadbd-safe-helper >/dev/null 2>&1; then
    alias mysqld_safe_helper="mariadbd-safe-helper"
fi
```

```bash
# Command name changes for MariaDB 11+
# Works everywhere if you source this file, but breaks pid="$!" number for mysqld

if command -v mariadb >/dev/null 2>&1; then
    mysql() { mariadb "$@"; }
fi

if command -v mariadb-access >/dev/null 2>&1; then
    mysqlaccess() { mariadb-access "$@"; }
fi

if command -v mariadb-admin >/dev/null 2>&1; then
    mysqladmin() { mariadb-admin "$@"; }
fi

if command -v mariadb-backup >/dev/null 2>&1; then
    mariabackup() { mariadb-backup "$@"; }
fi

if command -v mariadb-binlog >/dev/null 2>&1; then
    mysqlbinlog() { mariadb-binlog "$@"; }
fi

if command -v mariadb-check >/dev/null 2>&1; then
    mysqlanalyze() { mariadb-check "$@"; }
    mysqlcheck() { mariadb-check "$@"; }
    mysqloptimize() { mariadb-check "$@"; }
    mysqlrepair() { mariadb-check "$@"; }
fi

if command -v mariadb-convert-table-format >/dev/null 2>&1; then
    mysql_convert_table_format() { mariadb-convert-table-format "$@"; }
fi

if command -v mariadb-dump >/dev/null 2>&1; then
    mysqldump() { mariadb-dump "$@"; }
fi

if command -v mariadb-dumpslow >/dev/null 2>&1; then
    mysqldumpslow() { mariadb-dumpslow "$@"; }
fi

if command -v mariadb-find-rows >/dev/null 2>&1; then
    mysql_find_rows() { mariadb-find-rows "$@"; }
fi

if command -v mariadb-fix-extensions >/dev/null 2>&1; then
    mysql_fix_extensions() { mariadb-fix-extensions "$@"; }
fi

if command -v mariadb-hotcopy >/dev/null 2>&1; then
    mysqlhotcopy() { mariadb-hotcopy "$@"; }
fi

if command -v mariadb-import >/dev/null 2>&1; then
    mysqlimport() { mariadb-import "$@"; }
fi

if command -v mariadb-install-db >/dev/null 2>&1; then
    mysql_install_db() { mariadb-install-db "$@"; }
fi

if command -v mariadb-plugin >/dev/null 2>&1; then
    mysql_plugin() { mariadb-plugin "$@"; }
fi

if command -v mariadb-report >/dev/null 2>&1; then
    mysqlreport() { mariadb-report "$@"; }
fi

if command -v mariadb-secure-installation >/dev/null 2>&1; then
    mysql_secure_installation() { mariadb-secure-installation "$@"; }
fi

if command -v mariadb-setpermission >/dev/null 2>&1; then
    mysql_setpermission() { mariadb-setpermission "$@"; }
fi

if command -v mariadb-show >/dev/null 2>&1; then
    mysqlshow() { mariadb-show "$@"; }
fi

if command -v mariadb-slap >/dev/null 2>&1; then
    mysqlslap() { mariadb-slap "$@"; }
fi

if command -v mariadb-tzinfo-to-sql >/dev/null 2>&1; then
    mysql_tzinfo_to_sql() { mariadb-tzinfo-to-sql "$@"; }
fi

if command -v mariadb-upgrade >/dev/null 2>&1; then
    mysql_upgrade() { mariadb-upgrade "$@"; }
fi

if command -v mariadb-waitpid >/dev/null 2>&1; then
    mysql_waitpid() { mariadb-waitpid "$@"; }
fi

if command -v mariadbd >/dev/null 2>&1; then
    mysqld() { mariadbd "$@"; }
fi

if command -v mariadbd-multi >/dev/null 2>&1; then
    mysqld_multi() { mariadbd-multi "$@"; }
fi

if command -v mariadbd-safe >/dev/null 2>&1; then
    mysqld_safe() { mariadbd-safe "$@"; }
fi

if command -v mariadbd-safe-helper >/dev/null 2>&1; then
    mysqld_safe_helper() { mariadbd-safe-helper "$@"; }
fi
```

**Comment by rfay** [comment]
Thanks for exploring all this. It's annoying to us end-users that they care more about their branding than about our decades of finger-memory. 

We do need to either put your results somewhere else or retitle this or something so that we can find this when we need it. 

I guess the day may come that they force us to add `ddev mariadb` and such.

File: pkg/ddevapp/global_dotddev_assets/commands/db/mysql [hunk 2]
```diff
@@ -1,11 +1,17 @@
...
-## `ddev mysql --database=mysql -uroot -proot` gets you to the 'mysql' database with root privileges
+## Example: "ddev mysql" or "ddev mysql -udb -pdb" or "echo 'SHOW TABLES;' | ddev mysql"
+## `ddev mysql --database=mysql -udb -pdb` gets you to the 'mysql' database with 'db' user privileges
+## Aliases: mariadb
 ## DBTypes: mysql,mariadb
 ## ExecRaw: true
 
-mysql -udb -pdb --pager "$@"
+# Prefer mariadb instead of mysql if available
+if [[ "${DDEV_DATABASE}" == "mariadb"* ]] && command -v mariadb >/dev/null 2>&1; then
```
Comment (rfay) [inline pkg/ddevapp/global_dotddev_assets/commands/db/mysql:13]: Oh, you did it!

File: pkg/ddevapp/global_dotddev_assets/commands/db/mysqldump.example [hunk 3]
```diff
@@ -1,8 +1,16 @@
 #!/usr/bin/env bash
 
 ## #ddev-generated
-## Description: run mysqldump in web container
+## Description: run mysqldump or mariadb-dump in web container
```
Comment (rfay) [inline pkg/ddevapp/global_dotddev_assets/commands/db/mysqldump.example:4]: This is fine. I do wonder if this example (or the other examples) actually serve a purpose more than the docs.

File: docs/content/users/usage/commands.md [hunk 4]
```diff
@@ -1115,16 +1115,18 @@ ddev mutagen version
 
 ## `mysql`
 
-Run MySQL client in the database container (global shell db container command). This is only available on projects that use the `mysql` or `mariadb` database types.
+*Alias: `mariadb`.*
```
[the remark this answers is not in the document]
Comment (stasadev) [reply docs/content/users/usage/commands.md:1118]: The alias is good, but it doesn't have autocompletion in `ddev <TAB>`

I will move it to a new command (available only for `mariadb` database type), duplication in this case is justified.

File: docs/content/users/usage/commands.md [hunk 5]
```diff
@@ -1019,6 +1019,23 @@ Example:
 ddev mailpit
 ```
 
+## `mariadb`
+
+Run MariaDB client in the database container (global shell db container command). This is only available on projects that use the `mariadb` database type.
```
Comment (rfay) [inline docs/content/users/usage/commands.md:1024]:
```suggestion
Run MariaDB client in the database container (global shell db container command). This is only available on projects that use the `mariadb` database type. (This is the same as the `mysql` command, but MariaDB is changing the name of their CLI client.)
```

**Review by rfay — APPROVED** [review]
Works perfectly. Thanks for the thoughtful improvements.

**Comment by rfay** [comment]
I don't understand this failure, https://buildkite.com/ddev/ddev-macos-arm64-mutagen/builds/9528#01940865-77a7-45d9-8966-f33497bf2037/7967-8903

```
#11 [db 2/3] RUN getent group tty \|\| groupadd tty
  | #11 0.102 exec /bin/bash: exec format error
  | #11 ERROR: process "/bin/bash -c getent group tty \|\| groupadd tty" did not complete successfully: exit code: 1
 ```

"exec format error" usually means the binary can't be run for various reasons. In this case I think it would be complaining about `groupadd`. Or maybe it's complaining about `/bin/bash` itself.  This is on macstadium-m1, and it's Docker Desktop, so ARM64.  It's during the build of the dbserver. 

It's mariadb, 11.4. I went there and did a `docker pull ddev/ddev-dbserver-mariadb-11.4:20241223_stasadev_build_warn` and it was already up-to-date.

Docker Desktop had rosetta enabled on macstadium-m1. I disabled it.

Restarted test

**Comment by rfay** [comment]
Same failure in upstream/master, https://buildkite.com/ddev/ddev-macos-arm64-mutagen/builds/9524#01940877-6193-42d9-b769-622e30fb36db/7929-8866

Once again, it was on mariadb:11.4, but there it was against `ddev/ddev-dbserver-mariadb-11.4:20231216_bitnami_base_for_mysql` and here it's against `ddev/ddev-dbserver-mariadb-11.4:20241223_stasadev_build_warn`

Something is wrong with something.  It may be something wrong with our mariadb:11.4 build? (I do note that that build fails often due to timeouts, it seems to take longer than anything else and is the most likely thing to have to restart)

But the same test succeeded here on tb-macos-arm64-4: https://buildkite.com/ddev/ddev-macos-arm64-mutagen/builds/9528#019408f1-4827-4ed0-b944-42d398d996f6

Turning off rosetta on tb-macos-arm64-4 and re-running this.

**Comment by mikestreety** [comment]
Sorry to comment on an old MR, i've just spent some time debugging an issue which came back to this. Posting here for visibility more than anything requesting a change. Hoping it will help someone in the future as it is an unintended side-effect.

Our application (TYPO3) is using the `db`/`db` credentials to access the DB. We use deployer to pull down the database which relies on `MYSQL_PWD`. (https://github.com/sourcebroker/deployer-extended-database/blob/6431be49b95193c9fe6fb69ecb406ca3c5cf65d6/deployer/db/task/db_import.php#L55) to insert the credentials.

Something I learnt today is because the `.my.cnf` file exists with a password entry, this overrides any `MYSQL_PWD` variable set. With the change above, it was inserting the password as `root` alongside the `db` credentials.

**Comment by stasadev** [comment]
Hi @mikestreety,

Using `MYSQL_PWD` is confusing along with `.my.cnf`.

I would recommend to switch from `MYSQL_PWD=pass mysql -uuser` to `mysql -uuser -ppass` in your code.

Thanks for sharing your experience.

**Comment by rfay** [comment]
@mikestreety I found that baffling as well, looked at several ways around it. In the end though, it's pretty much a mistake to use .my.cnf and also $MYSQL_PWD. It had worked all these years because we were using db:db in the .my.cnf, which matched the $MYSQL_PWD. But having both of them i there was always a mistake.

**Comment by rfay** [comment]
@mikestreety I found that baffling as well, looked at several ways around it. In the end though, it's pretty much a mistake to use .my.cnf and also $MYSQL_PWD. It had worked all these years because we were using db:db in the .my.cnf, which matched the $MYSQL_PWD. But having both of them i there was always a mistake.

**Comment by mikestreety** [comment]
@stasadev @rfay yeah, I agree, I've never realised there was a `.my.cnf` file until this issue cropped up (or that it takes precedence over the `MYSQL_PWD` variable :) )

Will raise an issue with the deployer-extended team and see if it can be changed 👍  Thanks again!

**Pull request merged**